### PR TITLE
[vlive] New extractor for vlive.tv

### PR DESCRIPTION
**Pull request by ping**
"V" http://www.vlive.tv is a live broadcasting mobile platform targeted at South Korean celebrities

Similar to Periscope, videos are available for viewing after the live broadcast, but only on the app for now.

Example: http://m.vlive.tv/video/1326

### CODE REVIEW

**Comment by yan12125** [comment]
Why did you remove ```upload_date```?

File: youtube_dl/extractor/vlive.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        msgpad = '%.0f' % (time() * 1000)
+        md = b64encode(
+            hmac.new(self._SECRET.encode('ascii'),
+                     (url[:255] + msgpad).encode('ascii'), sha1).digest()
+        )
+        url += '&' + compat_urllib_parse.urlencode({'msgpad': msgpad, 'md': md})
+        playinfo = self._download_json(url, video_id, 'Downloading video json')
+
+        if playinfo.get('message', '') != 'success':
+            raise ExtractorError(playinfo['message'])
```
Comment (yan12125) [inline youtube_dl/extractor/vlive.py]: Here you should consider the case that ```playinfo``` does not have the ```'message'``` field, too.

File: youtube_dl/extractor/vlive.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        video_id = self._match_id(url)
+
+        webpage = self._download_webpage(
+            'http://m.vlive.tv/video/%s' % video_id,
+            video_id, note='Download video page')
+
+        title = self._og_search_title(webpage)
+        thumbnail = self._og_search_thumbnail(webpage)
+        creator = self._html_search_regex(
+            r'<span class="name">([^<>]+)</span>', webpage, 'creator')
```
Comment (yan12125) [inline youtube_dl/extractor/vlive.py]: ```<span[^>]+class="name">...``` is better.

File: youtube_dl/extractor/vlive.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        'url': 'http://m.vlive.tv/video/1326',
+        'md5': 'cc7314812855ce56de70a06a27314983',
+        'info_dict': {
+            'id': '1326',
+            'ext': 'mp4',
+            'title': '[V] Girl\'s Day\'s Broadcast',
+            'creator': 'Girl\'s Day',
+        },
+    }
+    _SECRET = 'rFkwZet6pqk1vQt6SxxUkAHX7YL3lmqzUMrU4IDusTo4jEBdtOhNfT4BYYAdArwH'
```
Comment (yan12125) [inline youtube_dl/extractor/vlive.py]: Is it possible to get the secret from the website rather using an hardcoded one? The key may be changed in the future and the extractor is broken before the new key is committed.

**Comment by ping** [comment]
@yan12125 ``upload_date`` was removed because they switched to displaying relative dates, like "10 minutes ago", "yesterday", "12 days ago", "1 month ago" instead of the yyyy.mm.dd format.

File: youtube_dl/extractor/vlive.py (hunk 2 above)
Comment (ping) [inline youtube_dl/extractor/vlive.py]: Changed.

File: youtube_dl/extractor/vlive.py (hunk 1 above)
Comment (ping) [inline youtube_dl/extractor/vlive.py]: Changed.

File: youtube_dl/extractor/vlive.py (hunk 3 above)
Comment (ping) [inline youtube_dl/extractor/vlive.py]: Not to my knowledge. At present, videos are only playable through the official ios/android apps and not the website.

File: youtube_dl/extractor/vlive.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+from ..utils import (
+    ExtractorError,
+    determine_ext
+)
+from ..compat import compat_urllib_parse
+
+
+class VLiveIE(InfoExtractor):
+    IE_NAME = 'vlive'
+    _VALID_URL = r'https?://(?:(www|m)\.)?vlive\.tv/video/(?P<id>[0-9]+)'
```
Comment (yan12125) [inline youtube_dl/extractor/vlive.py]: I've tried http://www.vlive.tv/video/1326 and got a 404 error. Do you have a www.vlive.tv example URL?

Comment (ping) [inline youtube_dl/extractor/vlive.py]: They used to redirect http://www.vlive.tv/video/1326 to http://m.live.tv/video/1326 for all requests, but it seems that they only do the redirect for mobile devices now (you can try with your phone, or toggle device mode in Chrome's Developer Tools). Not sure if this change in behaviour is a temporary bug.

When you share a link from the V app to Twitter, the link generated is still http://vlive.tv/video/1326 which redirects to http://www.vlive.tv/video/1326.

I'm inclined to leave www in the url regex for this.

Comment (yan12125) [inline youtube_dl/extractor/vlive.py]: Thanks for the information. Could you add some comments?

**Comment by ping** [comment]
@yan12125 Forgot to leave a comment that I've added a note about the mobile redirect in code. Not sure  if there's more you want documented. Let me know.

**Comment by yan12125** [comment]
Thanks for the contribution!

**Pull request merged**